1주차 미션 제출

jambottle opened the pull request:
# 1주차 미션 제출 (김재원)

## 구현 화면

<img width="1200" alt="week1_01" src="https://user-images.githubusercontent.com/72926450/148686832-862771f9-e85c-44bf-85a9-87f1606669da.png">
<img width="1200" alt="week1_02" src="https://user-images.githubusercontent.com/72926450/148686835-54198b8d-1856-4788-b93f-04ed70c0ac49.png">

## 접근 방법

* [x] 미션 코드를 구현하기 위해서 어떤 생각들을 했는지 정리해보세요.

**📌 개인적으로 1주차 미션에서 요구한 기능 구현 자체는 그리 어렵지 않았습니다.**
- v-model 디렉티브를 활용하여 input 태그 입력값의 양방향 데이터 바인딩 구현
- 버튼별 이벤트 발생에 따라 호출되어야 하는 문자열 회전 및 알림창 함수 구현

**📌 적절한 수준에서 컴포넌트 분리를 실시했습니다.**
- RotateText.vue 파일에 1주차 미션으로 구현한 기능들을 모아 하나의 단독 컴포넌트로 분리
- 알림창 컴포넌트만 Modal.vue 파일로 분리하였고, v-if 디렉티브와 slot으로 세부 기능 구현

**📌 Jest와 Vue Test Utils를 활용한 'TDD 방법론' 역시 처음 접해 다소 낯설기도 했지만, 이해했습니다.**
- Jest와 Vue Test Utils를 활용하여 앞서 구현한 세 가지 기능에 대한 테스트 코드 작성 (→ **테스트 대상 선정에 대해 재고**)
- **[리팩토링]**
  - ID 대신 data-test 속성을 선택자로 활용하는 방향으로 테스트 코드를 전반적으로 수정
  - 하위 컴포넌트까지 테스트 범위에 넣기 위해 shallowMount 대신 mount 메서드 활용
  - inputText(testText)를 테스트 코드 내에서 직접 설정하는 대신 setValue 메서드 활용
  - setValue 및 trigger의 비동기적 특성이 적용되는 테스트 코드에 async-await 구문 적용

**📌 코드 디버깅에 `alert`창이 자주 사용되므로 UI 측면에서 지양하는 편이 낫다는 것을 처음 알았습니다.**
- `alert` 함수로 알림창 구현 시, ESLint가 Custom UI의 구현으로 알림창을 대체할 것을 권고
- `// eslint-disable-next-line no-alert`를 추가하는 것만으로도 해결 가능하다는 것을 확인

## 특이사항

* [x] 구현하면서 궁금했던 점이나, 아쉬웠던 부분이 있으면 적어주세요.

𝑸. 그때그때 시각적으로 결과물을 확인할 수 있는 프론트엔드의 특성상 Unit Testing을 통해 요구사항의 만족 여부를 확인하는 것이 번거로운 일이 되지는 않을까요? 아직 제가 구현한 Unit Testing의 검증 수준이 그리 대단치 않아서(?) 더 그렇게 느끼는 것 같습니다.

> 강사님의 라이브 코딩을 통해서 **1)** 테스트 코드를 바탕으로 한 TDD 방법론 도입이 어떤 과정에서 진행되는지, **2)** 이런 방식으로 코드를 짜는 게 어떤 의미를 갖는지에 대해 처음 깨닫게 되었습니다. 다음 주차 미션에서는 TDD 방법론과 보다 친해질 수 있을 것 같습니다 😁

> **공유해주시는 배경지식과 공식문서를 먼저 숙지하는 습관을 들여야겠다**는 생각도 들었습니다! 충분히 숙지하지 않은 상태에서 코드를 작성하려고 하다 보니까 미션 출제에 담긴 의도를 파악하지 못한 채로 코드를 작성하게 되지 않았나 하는 아쉬움이 개인적으로 들었습니다.

## 제출 후 확인

* [x] 본인의 구현 결과를 `dev` 브랜치에 구현하고, `dev` 브랜치에서 `submission` 브랜치로 PR을 보내주세요.
* [x] Pull Request 생성 후 Reivewer에 @externship-master 와 @lkaybob 을 추가해주세요.

jambottle committed:
Fix: 쌍따옴표 자동 변환으로 인한 컴파일 에러 해결
- .eslintrc.js 파일 내 rules 부분에 쌍따옴표 관련 규칙 추가

jambottle committed:
Add: 1주차 미션 요구사항 구현
- v-model 디렉티브를 활용하여 input 태그 입력값의 양방향 데이터 바인딩 구현
- 버튼별 이벤트 발생에 따라 호출되어야 하는 문자열 회전 및 알림창 함수 구현

jambottle committed:
Modify: ESLint의 no-alert 경고에 대한 조치로 Modal 컴포넌트 분리
- alert 함수로 알림창 구현 시, ESLint가 Custom UI의 구현으로 알림창을 대체할 것을 권고
- Modal.vue 파일로 알림창 컴포넌트를 분리하고, v-if 디렉티브와 slot으로 세부 기능 구현

jambottle committed:
Test: 1주차 미션 요구사항 구현에 대한 컴포넌트 단위 테스트 실시
- Jest와 Vue Test Utils를 활용하여 앞서 구현한 세 가지 기능에 대한 테스트 코드 작성
- 각각의 케이스별로 테스트 내용을 구체적으로 명시하고, 이를 describe API로 그룹화
- 그 밖에도 jsconfign.js 파일에서 발생한 에러 해결을 위한 source-map 패키지 설치

jambottle committed:
Modify: RotateText 컴포넌트 분리

jambottle committed:
Remove: example.spec.js 파일 삭제

hogyun3709 commented on mission/tests/unit/RotateText.spec.js:
```diff
@@ -0,0 +1,45 @@
+    },
+  });
+
+  afterEach(() => {
```
안녕하세요 재원님, 그룹 5의 김호균입니다. 같은 그룹원은 아니지만 작성하신 TEST 코드가 인상 깊어 코멘트 남깁니다. 
기능 구현 요구사항을 모두 테스트 케이스로 구현할 수 있었다는 사실이 놀라웠습니다. 
더 나아가서 혹시 inputText 의  length 를 바탕으로 주어진 모든 문자열이 회전하는 테스트 상황을 가정 해보고 싶은데요.
테스트 케이스에서도 loop문을 활용하여 위 상황을 테스트 할 수있을지 궁금합니다.  
추가적으로 해당 테스트 코드를 저희그룹원들에게 소개해도 될지 여쭙고싶습니다.

jambottle replied:
호균님, 안녕하세요! 👋 다른 팀 멤버들의 코드까지 챙겨보시다니... 정말 대단하셔요 👍

제 테스트 코드는 당연히 공유하셔도 되구요, 오히려 소개해주셔서 영광입니다 ㅎㅎㅎ
아마 제 나름대로 주어진 요구사항들에 대해 최대한 테스트를 시도해보려고 노력한 덕분인 것 같은데요-

> 사실 저도 이번이 **TDD 방법론 도입 & 테스트 코드 작성**이 처음이다 보니까 뭔가 허전하다는 느낌이 있었어요...
> 자신이 없다 보니까 기능부터 먼저 구현한 뒤에 테스트 코드를 나중에 작성해서 더욱 부족했다는 아쉬움이 있었는데요 😢
> 
> 호균님 피드백을 들으니까 제 코드처럼 소위 `"Happy Case"`라고 불리는 일부 성공 케이스에 대해서만 검증하고 마는 것이 아니라 **'반복문, 조건문 등을 이용해서 발생할 수 있는 모든 상황을 커버할 수 있도록 테스트 코드를 짜야겠구나'** 하는 깨달음을 얻게 되네요!!

질문해주신 것처럼 저도 반복문을 활용해서 하나의 테스트 안에서 최대한 많은 케이스를 커버할 수 있을지 더 시도해봐야겠어요.
호균님도 혹시 추가로 아이디어 있으시면 언제든 댓글 남겨주세요, 저도 시도에 성공하면 코드 공유해드릴게요 :-D

hogyun3709 commented on mission/tests/unit/RotateText.spec.js:
```diff
@@ -0,0 +1,45 @@
+  });
+
+  afterEach(() => {
+    wrapper.vm.inputText = "ABLY Externship";
```
추가적으로 재원님의 테스트 코드 작성을 저희 그룹원에게 공유하여 나누고 싶은데 괜찮으실지 여쭙고 싶습니다.

allegoryj commented on mission/tests/unit/Home.spec.js:
```diff
@@ -0,0 +1,45 @@
+    wrapper.vm.clicks = 0;
+  });
+
+  it("Input에 값을 입력하면, Input에 입력한 값이 그 아래에 똑같이 표시된다.", () => {
```
테스트 코드 이름을 자세하게 작성해 주셔서 어떤 테스트를 진행할 것인지 이해하는데 큰 도움이 되었습니다.

allegoryj commented:
안녕하세요 재원님! 그룹 2 정진재입니다.
code presentation에서 기능들을 어떻게 구현하였는지, 테스트를 어떻게 진행하였는지 하나하나 자세하게 설명해주셔서 미션을 어떻게 진행하셨는지 이해하기 정말 수월하였습니다. 요구사항들을 정말 완벽하게 그리고 깔끔하게 구현하신 것 같아 배울 점이 많은 것 같습니다! 앞으로도 잘 부탁드립니다:))

allegoryj commented on mission/src/views/Home.vue:
```diff
@@ -3,40 +3,54 @@
 
 <script>
 import HelloWorld from "@/components/HelloWorld.vue";
+import Modal from "@/components/Modal.vue";
```
알림창 구현 시 발생하는 에러를 Modal 컴포넌트를 분리하여 기능 구현하시고 import하셔서 해결하신 점이 인상 깊었습니다!!

jambottle replied:
**UI 측면에서 `alert()` 함수로 구현되는 알림창은 디버깅할 때 자주 사용**되기 때문에 Custom UI로 새로 짜는 게 좋다더라고요-
저도 처음에는 당연한 듯이 `alert()` 함수를 썼었는데, 위 내용을 이번에 처음 알게 되어서 한번 적용해보았습니다 ㅎㅎㅎ

kyubumjang replied:
모달로 분리하여 기능을 구현하신 점에서 하나 배워갑니다! 제 github에 남겨주신 링크 잘 읽어보고 no-alert를 ESLint 룰을 수정하는 것이 아닌 Custom UI로 만들어보도록 하겠습니다! 감사합니다!

allegoryj commented on mission/tests/unit/Home.spec.js:
```diff
@@ -0,0 +1,45 @@
+    wrapper.vm.inputText = "ABLY Externship";
+    wrapper.vm.clicks = 0;
+  });
+
```
Test가 끝날 때마다 값을 초기화해주는 코드가 인상깊었습니다.

jambottle replied:
`describe API`로 여러 테스트 케이스를 묶어 보니까 **앞선 테스트의 결과가 다음 테스트로 그대로 전달**되는 걸 발견했어요!
만약 테스트의 의도가 초기값을 기준으로 하고 있다면, 테스트가 끝날 때마다 결과를 원상복구 시켜야 할 것 같아서 관련 기능을 찾아서 적용해보았습니다 🧐

christianB053 commented on mission/src/components/RotateText.vue:
```diff
@@ -0,0 +1,64 @@
+    rotateText() {
+      const str1 = this.inputText.slice(1);
+      const str2 = this.inputText.slice(0, 1);
+      this.inputText = str1 + str2;
```
문자열 회전 시키기를 구현하지 못했었는데 재원님이 하신 것처럼 slice 메서드를 각각 변수로 지정해서
this.inputText = str1 + str2로 구현할 수 있다는 것을 알게 되어 따라해 봤더니 문자열 회전시키기 성공했어요!
slice 메서드를 책으로만 접해 봐서 찾아 보면서도 이해가 잘 되지 않았었는데
slice(1)은 arr[1]이후 모든 요소 복사하여 반환(rojectLion)
slice(0,1)은 arr[0]부터 arr[1]이전까지 복사하여 반환(P)이기 떄문에
str1 + str2 = rojectLionP이 되는 것이네요!

christianB053 commented on mission/src/components/RotateText.vue:
```diff
@@ -0,0 +1,64 @@
+    },
+    showModal() {
+      this.isModalShown = true;
+      this.clicks += 1;
```
알림창은 모달 컴포넌트로 구현할 수 있다는 것을 알게 되어
이 방법을 사용해서 다시 시도해 보도록 하겠습니다!

Soohyeon02 replied:
경민님도 그렇고, 재원님도 그렇고 모달을 사용하셨는데 저도 다음에 한 번 사용해봐야겠습니다.
좋은 방법을 배워갑니다!

kyubumjang replied:
모달을 사용하는 방식이 alert를 사용하지 않고 구현하는 좋은 방법인 것 같습니다!

christianB053 reviewed (commented):
안녕하세요. 재원님, 2조 김현빈입니다!
코딩 공부를 막 시작한 초보자로 도움되는 리뷰를 하지 못한 점 양해 부탁드립니다.
TDD는 시도도 못 해 봤는데 작성하신 것 참고하여 공부해 보겠습니다.
남은 미션도 화이팅하세요! :)

daaeaad commented on mission/src/components/RotateText.vue:
```diff
@@ -0,0 +1,64 @@
+      this.inputText = str1 + str2;
+    },
+    showModal() {
+      this.isModalShown = true;
```
예시로 아래 내용과 같은 모달창을 열고 닫는 함수를 따로 만들어서 사용했어도 좋았을 것 같다는 생각이 들었습니다!
```
toggleModal() {
  this.isModalShown = !this.isModalShown;
}
```

이미 생각해보셨던 내용일 수 있지만 조심스레 의견 남겨봅니다.
혹시 따로 함수를 만들지 않으신 이유는 이번 프로젝트에서는 모달을 한번만 사용하기 때문이었을까요?  😶

jambottle replied:
안녕하세요, 다애님! 👋 (결코 작지 않은!!) 의견과 질문 남겨주셔서 감사합니다-

제게 대단한 의도가 있었던 것은 아니지만, 다애님의 의견과 다른 방식으로 함수를 구현한 가장 결정적인 이유를 찾아보자면...
이번에 `showModal` 함수를 만들면서 아래와 같은 두 가지 역할을 한 번에 담기로 결정했기 때문인 것 같아요. 🤔

1. 이 함수가 실행되면, 모달창이 열려야 한다. (⇒ `isModalShown`이 **true**가 되어야 한다.)
2. 모달창이 열리는 동시에 클릭 횟수가 증가해야 한다. (⇒ `clicks`가 **1씩 증가**해야 한다.)

> 남겨주신 의견대로 **toggle 방식을 채택할 수도 있었겠지만, 그렇게 되면 클릭 횟수 증가 기능을 같은 함수에 담지 못하게 될 것**이므로 toggle 방식을 미리 알았더라도 지금의 저는 위와 같은 방식을 선택하지 않았을까 싶습니다. (모달창이 닫힐 때에도 클릭 횟수가 증가하면, 주어진 요구사항과 맞지 않기 때문이에요.)

물론, 클릭 횟수를 증가시키는 기능을 별도의 함수로 분리해도 되긴 할 텐데요... 사실 이번에 `Modal.vue` 컴포넌트 파일을 만들 때에 **[Vue.js Docs의 예제 코드](https://v3.ko.vuejs.org/examples/modal.html#modal-component)를 그대로 복사해** 가져다 쓰면서 이렇게 구현하게 된 것도 없지 않아 있습니다 😅;;

jambottle replied:
간단히 정리하자면, 저는 아래와 같이 선택했다고 봐주시면 될 것 같습니다!!

- ✅ `모달창을 열면서 클릭 횟수를 증가시키는 함수` & `모달창을 닫는 함수`로 분리
- ❎ `모달창을 열고 닫는 함수` & `클릭 횟수를 증가시키는 함수`로 분리

혹시 다애님도 더 구체적인 의견이 있으시다면, 댓글로 남겨주세요 :-D
좋은 질문 주셔서 다시 한번 감사드립니다 🙏

daaeaad reviewed (commented):
안녕하세요, 그룹 4의 김다애 입니다!
같은 그룹원은 아니지만 재원님의 테스트 코드가 참고하기 좋다는 소문을 듣고 왔습니다  👀
테스트를 잘 진행해보려고 공식문서나 유튜브 영상을 많이 찾아보았는데요,
같은 내용의 프로젝트를 테스트하셔서 그런진 몰라도 재원님 코드 프레젠테이션 영상을 보고 
지금까지 참고한 자료들 보다 테스트 코드 작성하는 방법을 이해하는 데에 많이 도움되었습니다!
대단히 도움드릴 만한 코멘트는 남기지 못하지만 작은 의견과 질문 하나 남기고 갑니다 =3
좋은 자료 잘 보고 갑니다 :)

ddukddak-kim commented:
재원님 안녕하세요 :) 
이번 익스텐션 기간동안 Peer 리뷰를 함께 할 2 그룹 김경민 입니다. 
만나서 반가워요~ 😎

이번주 미션 수행하느라 고생 많으셨습니다-!

Soohyeon02 commented:
안녕하세요. 이번에 같은 조로 배정된 황수현입니다!
저는 TDD를 적용하지 못하여서 그 부분을 유심하게 보았는데, 재원님 코드를 보면서 도움이 많이 되었습니다.
미션 수행하느라 수고하셨습니다!

kyubumjang commented on mission/src/views/Home.vue:
```diff
@@ -3,40 +3,54 @@
+    rotateText() {
+      const str1 = this.inputText.slice(1);
+      const str2 = this.inputText.slice(0, 1);
+      this.inputText = str1 + str2;
```
변수명도 신경을 쓰셨군요! 변수명을 명료하고 이해할 수 있게 정하는게 어렵네요..😂

kyubumjang commented on mission/tests/unit/Home.spec.js:
```diff
@@ -0,0 +1,45 @@
+    expect(wrapper.vm.isModalShown).toBeTruthy();
+    expect(wrapper.vm.clicks).toBe(3);
+  });
+});
```
테스트 코드 보고 많이 배웠습니다. 어떤 식으로 해야할 지 잘 몰랐는데 아주 좋은 예시인 것 같습니다. 다른 jest를 사용한 테스트 코드들도 봤는데 도저히 모르겠어서 mount나 shallowMount 없이 상황을 가정하고 그에 대한 결과만 나오도록 했습니다. 기능 구현에 초점을 뒀었는데 이런 방식을 사용해서 Test를 진행하고 구현을 하게되면 한번이라도 더 요구사항에 대해 생각해보고 불확실성을 줄여나갈 수 있겠네요.👍👍

KyubumJang commented:
재원님 안녕하세요! 같은 그룹2로 배정된 장규범입니다!

Code를 보고 많은 것을 얻어갑니다. 동기부여도 많이 되구요! 테스트 코드나 실제 구현하시는 과정, 코드를 짤 때 어떤 방식으로 구현 하는지 자세하게 말씀해주셔서 아주 좋았습니다. 댓글 정성스럽게 하나하나 알아보시고 링크까지 해주시는 점도 정말 감사드립니다. 이번주 수고 많으셨습니다!👍

앞으로도 잘 부탁드립니다.😁

lkaybob commented on mission/tests/unit/RotateText.spec.js:
```diff
@@ -0,0 +1,45 @@
+  const wrapper = shallowMount(RotateText, {
+    data() {
+      return {
+        inputText: "ABLY Externship",
```
inputText를 직접 설정하는 것은 input 자체를 테스트하는 것이 아닙니다. setValue 등을 이용해서 테스트해보세요.

https://next.vue-test-utils.vuejs.org/guide/essentials/a-crash-course.html#adding-a-new-todo

jambottle replied:
아직 TDD 방법론에 대한 이해도 부족하고, Jest 활용법에 대해서도 많이 미숙했던 것 같습니다...
우선, 말씀해주신 대로 `data`를 테스트 코드 내에서 직접 설정하는 것 대신 `setValue` 메서드로 대체해보겠습니다! 🙏

lkaybob commented on mission/tests/unit/RotateText.spec.js:
```diff
@@ -0,0 +1,45 @@
+  it("'제출'을 1회 누르면, Input에 입력한 값과 아래에 표시된 문자열이 왼쪽으로 1칸씩 회전한다.", () => {
+    const rotateButton = wrapper.find("#rotate");
+    rotateButton.trigger("click");
+    expect(wrapper.vm.inputText).toBe("BLY ExternshipA");
```
구현 내부에 직접 접근하는 것은 지양해주세요
* wrapper에서 찾은 node의 특정 값을 찾고 싶으시면 `wrapper.find('#output').text()`와 같이 찾아주세요.

https://next.vue-test-utils.vuejs.org/guide/essentials/easy-to-test.html#do-not-test-implementation-details

jambottle replied:
𝑸. 컴포넌트 구현 내부에 선언된 `data`가 어떻게 변화하는지를 추적해보려고 하다 보니 이렇게 테스트 코드를 작성하게 된 것 같아요!
혹시 node의 특정 값을 찾는 것 대신, `data`의 변화에 대해서 직접 테스트하고 싶은 경우에는 어떻게 해야 할까요? 🤔

lkaybob reviewed (commented):
* 전반적으로 직관적인 메시지들(커밋 메시지, identifier들)을 남겨주셔서 이해하기 편했습니다.
* 자잘한 변경사항들이 많이 들어간 것 같습니다. (예. public/index.html)
  * 자잘한 변경사항이 많이 들어가면 리뷰 상대방이 유의미한 변경사항을 찾는데 어려울 수 있습니다.
  * rule 변경으로 인한 변경사항들(예. single quote)도 존재하는 것 같은데, 협업하는 상황에서 갑작스런 변경은 지양해주셨으면 합니다.
* 테스트 대상의 선정이 조금 아쉽습니다. [Crash Course](https://next.vue-test-utils.vuejs.org/guide/essentials/a-crash-course.html) 문서와 내일 라이브세션의 내용을 참고해주세요.

jambottle committed:
Refactor: 쌍따옴표 자동 변환으로 인한 컴파일 에러 해결 복구
- .eslintrc.js 파일 내 rules 부분에 추가했던 쌍따옴표 관련 규칙을 삭제
- VSCode 기본 설정에서 defaultFormatter를 ESLint로 대체하여 해결

jambottle committed:
Refactor: RotateText.spec.js 테스트 코드 1차 리팩토링
- trigger의 비동기적 특성이 적용되는 테스트 코드에 async-await 구문 적용
- inputText를 테스트 코드 내에서 직접 설정하는 대신 setValue 메서드 활용

jambottle committed:
Refactor: RotateText.spec.js 테스트 코드 2차 리팩토링
- ID 대신 data-test 속성을 선택자로 활용하는 방향으로 테스트 코드를 전반적으로 수정
- 하위 컴포넌트까지 테스트 범위에 넣기 위해 shallowMount 대신 mount 메서드 활용
